csclient/csclient.go

```diff
@@ -156,6 +158,126 @@ func (c *Client) GetArchive(id *charm.URL) (r io.ReadCloser, eid *charm.URL, has
 	return resp.Body, eid, hash, resp.ContentLength, nil
 }
 
+// ListResources retrieves the metadata about resources for the given charm.
+func (c *Client) ListResources(ids ...*charm.URL) (map[string][]params.Resource, error) {
+	// Prepare the request to the charm store.
+	urls := make([]string, len(ids))
+	values := url.Values{}
+	for i, id := range ids {
+		url := id.WithRevision(-1).String()
+		urls[i] = url
+		values.Add("id", url)
+	}
+	u := url.URL{
+		Path:     "/meta/resources",
+		RawQuery: values.Encode(),
+	}
+
+	// Execute the request and retrieve results.
+	var results map[string][]params.Resource
+
+	if err := c.Get(u.String(), &results); err != nil {
+		return nil, errgo.NoteMask(err, "cannot get resource metadata from the charm store", errgo.Any)
+	}
+
+	return results, nil
+}
+
+// UploadResource uploads the bytes for a resource.
+func (c *Client) UploadResource(id *charm.URL, name, path string) (revision int, err error) {
+	file, err := os.Open(path)
+	if err != nil {
+		return -1, errgo.Mask(err)
+	}
+	defer file.Close()
+	hash, size, err := readerHashAndSize(file)
+	if err != nil {
+		return -1, errgo.Mask(err)
+	}
+
+	// Prepare the request.
+	req, err := http.NewRequest("PUT", "", nil)
+	if err != nil {
+		return -1, errgo.Notef(err, "cannot make new request")
+	}
+	req.Header.Set("Content-Type", "application/octet-stream")
+	req.ContentLength = size
+
+	hash = url.QueryEscape(hash)
+	path = url.QueryEscape(path)
+
+	url := fmt.Sprintf("/%s/resources/%s?hash=%s&filename=%s", id.Path(), name, hash, path)
+
+	// Send the request.
+	resp, err := c.DoWithBody(req, url, file)
+	if err != nil {
+		return -1, errgo.NoteMask(err, "cannot post resource", errgo.Any)
+	}
+	defer resp.Body.Close()
+
+	// Parse the response.
+	var result params.ResourceUploadResponse
+	if err := parseResponseBody(resp.Body, &result); err != nil {
+		return -1, errgo.Mask(err)
+	}
+	return result.Revision, nil
+}
+
+// ResourceData
+type ResourceData struct {
+	io.ReadCloser
+	Revision int
+	Hash     string
+	Size     int64
+}
+
+// GetResource retrieves the archive for the given charm or bundle, returning a
+// reader its data can be read from, the fully qualified id of the
+// corresponding entity, the SHA384 hash of the data and its size.
+func (c *Client) GetResource(id *charm.URL, name string) (result ResourceData, err error) {
```

Comment:
revision must be passed in rather than returned (in ResourceData).  If revision < 0 then we use the latest one.